### PR TITLE
 [Portfolio] use decimal.Decimal instead of floats

### PR DESCRIPTION
**Pull request by GuillaumeDSM**

### CODE REVIEW

File: octobot_trading/personal_data/portfolios/__init__.py [hunk 1]
```diff
@@ -58,6 +59,10 @@
     MarginPortfolio,
     SpotPortfolio,
 )
+from octobot_trading.personal_data.portfolios.portfolio_util import (
```
Comment (Herklos) [inline octobot_trading/personal_data/portfolios/__init__.py:62]: You forgot to also add it to `__init__.pxd`

File: octobot_trading/constants.py [hunk 2]
```diff
@@ -42,6 +43,11 @@
 DEFAULT_BACKTESTING_TIME_LAG = 0
 INFINITE_MAX_HANDLED_PAIRS_WITH_TIMEFRAME = -1
 
+# Decimal default values (decimals are immutable, can be stored is constant)
+ZERO = decimal.Decimal(0)
```
Comment (Herklos) [inline octobot_trading/constants.py:47]: :+1:

File: octobot_trading/exchanges/implementations/spot_ccxt_exchange.py [hunk 3]
```diff
@@ -229,7 +230,7 @@ async def switch_to_account(self, account_type: enums.AccountTypes):
         return await self.connector.switch_to_account(account_type=account_type)
 
     def parse_balance(self, balance):
-        return self.connector.parse_balance(balance)
+        return personal_data.parse_decimal_portfolio(self.connector.parse_balance(balance))
```
Comment (Herklos) [inline octobot_trading/exchanges/implementations/spot_ccxt_exchange.py:233]: :+1:

File: octobot_trading/personal_data/orders/order.py [hunk 4]
```diff
@@ -98,8 +100,9 @@ def get_logger_name(self):
         return self.logger_name
 
     def update(self, symbol, order_id="", status=enums.OrderStatus.OPEN,
-               current_price=0.0, quantity=0.0, price=0.0, stop_price=0.0,
-               quantity_filled=0.0, filled_price=0.0, average_price=0.0, fee=None, total_cost=0.0,
+               current_price=constants.ZERO, quantity=constants.ZERO, price=constants.ZERO, stop_price=constants.ZERO,
```
Comment (Herklos) [inline octobot_trading/personal_data/orders/order.py:103]: :+1:

File: octobot_trading/personal_data/orders/order_util.py [hunk 5]
```diff
@@ -107,11 +110,11 @@ async def get_pre_order_data(exchange_manager, symbol: str, timeout: int = None)
     current_market_quantity = exchange_manager.exchange_personal_data.portfolio_manager.portfolio\
         .get_currency_portfolio(market)
 
-    market_quantity = current_market_quantity / mark_price if mark_price else 0
+    market_quantity = current_market_quantity / decimal_mark_price if decimal_mark_price else 0
```
Comment (Herklos) [inline octobot_trading/personal_data/orders/order_util.py:113]: Shouldn't it be a Decimal.Zero ?

File: octobot_trading/personal_data/orders/types/limit/limit_order.py [hunk 6]
```diff
@@ -37,6 +37,7 @@ async def update_order_status(self, force_refresh=False):
 
     async def wait_for_price_hit(self):
         await asyncio.wait_for(self.limit_price_hit_event.wait(), timeout=None)
+        # negative pf here
```
Comment (Herklos) [inline octobot_trading/personal_data/orders/types/limit/limit_order.py:40]: ?

File: octobot_trading/personal_data/portfolios/portfolio_util.py [hunk 7]
```diff
@@ -28,6 +28,13 @@ def parse_decimal_portfolio(portfolio):
     return decimal_portfolio
 
 
+def parse_decimal_config_portfolio(portfolio):
+    decimal_portfolio = {}
+    for symbol, symbol_balance in portfolio.items():
+        decimal_portfolio[symbol] = decimal.Decimal(str(symbol_balance))
+    return decimal_portfolio
```
Comment (Herklos) [inline octobot_trading/personal_data/portfolios/portfolio_util.py:35]:
```
return {
    symbol: decimal.Decimal(str(symbol_balance))
    for symbol, symbol_balance in portfolio.items()
}
```

File: octobot_trading/personal_data/orders/order_util.py (hunk 5 above)
Comment (GuillaumeDSM) [reply octobot_trading/personal_data/orders/order_util.py:113]: yes !

File: octobot_trading/personal_data/orders/types/limit/limit_order.py (hunk 6 above)
Comment (GuillaumeDSM) [reply octobot_trading/personal_data/orders/types/limit/limit_order.py:40]: woops ^^

**Review by Herklos — APPROVED** [review]

**Review by Herklos — COMMENTED** [review]
:+1: nice work!

**Review by Herklos — APPROVED** [review]

**Pull request merged**